pkg/metrics/check_metrics.go

```diff
@@ -0,0 +1,158 @@
+	tlmCheckMetricsWaiting = telemetry.NewGauge("check_metrics", "waiting_metrics", []string{"stateful"}, "Gauge of expired check metrics waiting for timeout")
+	tlmCheckMetricsAdded   = telemetry.NewCounter("check_metrics", "created_total", []string{"stateful"}, "Count of new check metrics added")
+	tlmCheckMetricsExpired = telemetry.NewCounter("check_metrics", "expired_total", []string{"stateful"}, "Count of expired metrics")
+	tlmCheckMetricsRemoved = telemetry.NewCounter("check_metrics", "removed_total", []string{"stateful"}, "Count of removed metrics")
+
+	checkMetricsAddSampleTelemetry = &AddSampleTelemetry{
+		Total:     tlmCheckMetricsAdded.WithValues("sum"),
+		Stateful:  tlmCheckMetricsAdded.WithValues("true"),
+		Stateless: tlmCheckMetricsAdded.WithValues("false"),
+	}
+	tlmCheckMetricsExpiredTotal     = tlmCheckMetricsExpired.WithValues("sum")
+	tlmCheckMetricsExpiredStateful  = tlmCheckMetricsExpired.WithValues("true")
+	tlmCheckMetricsExpiredStateless = tlmCheckMetricsExpired.WithValues("false")
+	tlmCheckMetricsRemovedTotal     = tlmCheckMetricsRemoved.WithValues("sum")
+	tlmCheckMetricsRemovedStateful  = tlmCheckMetricsRemoved.WithValues("true")
+	tlmCheckMetricsRemovedStateless = tlmCheckMetricsRemoved.WithValues("false")
+)
+
+// CheckMetrics stores metrics for the check sampler.
+//
+// This is similar to ContextMetrics, but provides additional facility to remove metrics.
+type CheckMetrics struct {
+	expireMetrics bool
+	// additional time to keep stateful metrics in memory, after the context key has expired
+	statefulTimeout float64
+	metrics         ContextMetrics
+	deadlines       map[ckey.ContextKey]float64
+}
+
+// NewCheckMetrics returns new CheckMetrics instance.
+func NewCheckMetrics(expireMetrics bool, statefulTimeout time.Duration) CheckMetrics {
+	return CheckMetrics{
+		expireMetrics:   expireMetrics,
+		statefulTimeout: statefulTimeout.Seconds(),
+		metrics:         MakeContextMetrics(),
+		deadlines:       nil,
+	}
+}
+
+// AddSample adds a new sample to the metric with contextKey, initializing a new metric as necessary.
+//
+// If contextKey is scheduled for removal (see Remove), it will be unscheduled.
+//
+// See also ContextMetrics.AddSample().
+func (cm *CheckMetrics) AddSample(contextKey ckey.ContextKey, sample *MetricSample, timestamp float64, interval int64) error {
+	if cm.deadlines != nil {
+		delete(cm.deadlines, contextKey)
+	}
+	return cm.metrics.AddSample(contextKey, sample, timestamp, interval, checkMetricsAddSampleTelemetry)
+}
+
+// Remove deletes metrics associated with the given contextKeys.
+//
+// Some metrics can not be removed as soon as we stop receiving them (see
+// Metric.isStateful()). Stateful metrics will be kept around additional `cm.statefulTimeout`
+// time after this call, before ultimately removed. Call to AddSample will cancel delayed
+// removal.
+func (cm *CheckMetrics) Remove(contextKeys []ckey.ContextKey, timestamp float64) {
```

Comment:
Maybe this method should be called "expire", since it _expires_ everything but only _removes_ some?